google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -381,32 +388,70 @@ private static boolean isWindows() {
-    try {
-      Class<?> factoryClass =
-          Class.forName("com.google.appengine.api.appidentity.AppIdentityServiceFactory");
-      Class<?> serviceClass =
-          Class.forName("com.google.appengine.api.appidentity.AppIdentityService");
-      Method method = factoryClass.getMethod("getAppIdentityService");
-      Object appIdentityService = method.invoke(null);
-      method = serviceClass.getMethod("getServiceAccountName");
-      String serviceAccountName = (String) method.invoke(appIdentityService);
-      int indexOfAtSign = serviceAccountName.indexOf('@');
-      return serviceAccountName.substring(0, indexOfAtSign);
-    } catch (ClassNotFoundException exception) {
-      if (System.getProperty("com.google.appengine.runtime.version") != null) {
-        // Could not resolve appengine classes under GAE environment.
-        throw new RuntimeException("Google App Engine runtime detected "
-            + "(the environment variable \"com.google.appengine.runtime.version\" is set), "
-            + "but unable to resolve appengine-sdk classes. "
-            + "For more details see "
-            + "https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/APPENGINE.md");
+    if (PlatformInformation.isOnGAEStandard7()) {
+      try {
+        Class<?> factoryClass =
+                Class.forName("com.google.appengine.api.appidentity.AppIdentityServiceFactory");
+        Class<?> serviceClass =
+                Class.forName("com.google.appengine.api.appidentity.AppIdentityService");
+        Method method = factoryClass.getMethod("getAppIdentityService");
+        Object appIdentityService = method.invoke(null);
+        method = serviceClass.getMethod("getServiceAccountName");
+        String serviceAccountName = (String) method.invoke(appIdentityService);
+        int indexOfAtSign = serviceAccountName.indexOf('@');
+        return serviceAccountName.substring(0, indexOfAtSign);
+      } catch (ClassNotFoundException exception) {
+        if (System.getProperty("com.google.appengine.runtime.version") != null) {
+          // Could not resolve appengine classes under GAE environment.
+          throw new RuntimeException("Google App Engine runtime detected "
+                  + "(the environment variable \"com.google.appengine.runtime.version\" is set), "
+                  + "but unable to resolve appengine-sdk classes. "
+                  + "For more details see "
+                  + "https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/APPENGINE.md");
+        }
+        return null;
+      } catch (Exception ignore) {
+        return null;
       }
-      return null;
-    } catch (Exception ignore) {
-      return null;
+    } else {
+      //for GAE flex and standard Java 8 environment
+      String projectId = System.getenv("GCLOUD_PROJECT");
+      if (projectId == null) {
+        projectId = System.getenv("GOOGLE_CLOUD_PROJECT");
+      }
+      if (projectId == null) {
```

Comment:
I suggested the Metadata server, but really you can get it from GAE_APPLICATION on J8 standard and remove the prefix anything before & including the **~**

So the order becomes:
GCLOUD_PROJECT
GOOGLE_CLOUD_PROJECT
getAppEngineAppId()  i.e. System.getProperty("com.google.appengine.application.id"); 
GAE_APPLICATION (removing the prefix)

We need to document that users should set GOOGLE_CLOUD_PROJECT when running locally, until our tooling sets "com.google.appengine.application.id"